File: max-consecutive-ones-iii.cc
// Max Consecutive Ones III
class Solution {
public:
  int longestOnes(vector<int>& A, int K) {
    int n = A.size(), r = 0;
    for (int i = 0, j = 0; j < n; j++) {
      if (!A[j] && --K < 0)
        while (A[i++]);
      r = max(r, j-i+1);
    }
    return r;
  }
};

File: max-consecutive-ones-iii.nim
proc longestOnes(nums: ptr UncheckedArray[cint], numsSize: int, k: cint): int {.exportc.} =
  var
    l = 0
    bal = cast[int](k)
  for r in 0..<numsSize:
    if cast[int](nums[r]) == 0:
      dec bal
      if bal < 0:
        while (inc l; nums[l-1] > 0): discard
    result = max(result, r-l+1)
